rt_gene/src/rt_bene/estimate_blink_pytorch.py

```diff
@@ -0,0 +1,48 @@
+#! /usr/bin/env python
+
+from tqdm import tqdm
+from rt_gene.download_tools import download_blink_models
+from rt_bene.estimate_blink_base import BlinkEstimatorBase
+from rt_bene.blink_estimation_models_pytorch import BlinkEstimationModelResnet18
+import os
+import cv2
+import torch
+from torchvision import transforms
+
+
+class BlinkEstimatorPytorch(BlinkEstimatorBase):
+
+    def __init__(self, device_id_blink, model_files, threshold):
+        super(BlinkEstimatorPytorch, self).__init__(device_id=device_id_blink, threshold=threshold)
+        download_blink_models()
+        if "OMP_NUM_THREADS" not in os.environ:
+            os.environ["OMP_NUM_THREADS"] = "8"
+        tqdm.write("PyTorch using {} threads.".format(os.environ["OMP_NUM_THREADS"]))
+
+        self._transform = transforms.Compose([lambda x: cv2.resize(x, dsize=(224, 224), interpolation=cv2.INTER_CUBIC),
+                                              transforms.ToTensor(),
+                                              transforms.Normalize(mean=[0.485, 0.456, 0.406],
+                                                                   std=[0.229, 0.224, 0.225])])
+
+        self._models = []
+        for ckpt in model_files:
+            _model = BlinkEstimationModelResnet18()
+            _model.load_state_dict(torch.load(ckpt))
+            _model.to(self.device_id)
+            _model.eval()
+            self._models.append(_model)
+
+        tqdm.write('Loaded ' + str(len(self._models)) + ' model(s)')
+        tqdm.write('Ready')
+
+    def predict(self, left_eyes, right_eyes):
+        transformed_left = torch.stack(left_eyes).to(self.device_id)
+        transformed_right = torch.stack(right_eyes).to(self.device_id)
+
+        result = [torch.sigmoid(model(transformed_left, transformed_right)).detach().cpu() for model in self._models]
+        result = torch.stack(result, dim=1)
+        result = torch.mean(result, dim=1).numpy()
+        return result
+
+    def inputs_from_images(self, left, right):
```

Comment:
No flipping needed here? Elsewhere you flip the images, don't you?

Reply:
The tensorflow training requires the right image to be flipped; but I didn't see a particular reason for it so the dataset generation, training and inference do not flip in the pytorch version...
I could do it to maintain consistency between the two versions but I personally feel that we don't need it.

Reply:
Sounds good to me, thanks for clarifying :)

Reply:
By the way, flipping was done originally for weight sharing - I can't remember whether we ended up using it. In this case the idea is that the two images are more similar if they're flipped.

Reply:
Ah that makes sense - in the TF version, there isn't any particular weight sharing going on so I don't think it is being used. There's two independent neural networks that do inference on each eye...much like RT-GENE.